### PR TITLE
Run database cleanup in the background

### PR DESCRIPTION
**Pull request by mplorentz**
## Issues covered
#1426 

## Description
This moves our call to run the database cleanup routine from app launch to a background task. Running the cleanup on app startup causes UI bugs because the views sometimes fetch Core Data objects and display them, then those objects are deleted, and then even after we download them again SwiftUI does not properly observe that they are back. 

I tried waiting to initialize the UI until the cleanup has finished, but the cleanup can take several seconds in some cases, so that increased app startup time too much.

I also tried just deleting all Events to make the cleanup routine faster so we could run it at startup. But that got a lot of pushback on how that would break offline support, and I would need to re-design our code to republish events that failed to publish the first time. So I tabled that too.

I also tried running the cleanup in applicationWillTerminate, but the app is terminated by the OS before the cleanup can finish.

So I landed on running a BGAppProcessingTask, which in my experience gets run about once a day usually while the phone is charging. The downside to this approach is that if someone for some reason downloads a ton of Nostr events in a single session their database will get huge and the app will get slow, and restarting the app will not run the cleanup and fix it. But I think this case will be quite rare. On iOS it's very hard to keep the app open for many hours. On macOS it's easy but we have good Posthog stats on database sizes and most user's databases are staying quite small. I think it's something we should keep an eye on.

## How to test
Unfortunately Apple seems to have broken the old method of manually triggering BGAppProcessingTasks from the debugger, and they didn't replace it with anything. I have been testing like this:
1. Build and run the app on a real device
2. Put the app into the background
3. Wait 24 hours
4. Launch the app and export the logs
5. Filter the logs for "analytics" to see if the database cleanup routine was run.

If you don't want to wait 24 hours I have attached my [logs](https://github.com/user-attachments/files/17246090/com.verse.Nos-dev.2024-10-02--16-01-38-106.log) from the past 24 hours. And here are the relevant lines showing that it ran at 2:30am and 3am for me last night:
```
2024/10/02 16:09:10:392  LOG:INFO: Analytics: Identified npub1sydml5ghw26mnaglpf99esnmj7pa7vmhk4lvfydlz5cqnks8pnfsq4vsz8
2024/10/02 16:09:10:409  LOG:INFO: Analytics: Home Tab Tapped
2024/10/03 02:30:02:951  LOG:INFO: Analytics: Identified npub1sydml5ghw26mnaglpf99esnmj7pa7vmhk4lvfydlz5cqnks8pnfsq4vsz8
2024/10/03 02:30:03:437  LOG:INFO: Analytics: Database Cleanup Started
2024/10/03 02:30:03:832  LOG:INFO: Analytics: Identified npub1sydml5ghw26mnaglpf99esnmj7pa7vmhk4lvfydlz5cqnks8pnfsq4vsz8
2024/10/03 02:30:03:893  LOG:INFO: Analytics: Home Tab Tapped
2024/10/03 02:30:06:781  LOG:INFO: Analytics: Database Statistics: ["NosNotification": 16, "Follow": 59869, "AuthorReference": 1272, "Event": 3739, "EventReference": 1131, "Relay": 528, "Author": 14353]
2024/10/03 02:30:06:781  LOG:INFO: Analytics: Database Cleanup Completed: ["duration": 3.3444979190826416]
2024/10/03 03:00:11:515  LOG:INFO: Analytics: Database Cleanup Started
2024/10/03 03:00:12:784  LOG:INFO: Analytics: Database Statistics: ["AuthorReference": 1205, "Author": 14353, "Follow": 59869, "EventReference": 1074, "NosNotification": 16, "Event": 3688, "Relay": 528]
2024/10/03 03:00:12:784  LOG:INFO: Analytics: Database Cleanup Completed: ["duration": 1.2699642181396484]

```

### CODE REVIEW

File: Nos.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -2863,8 +2863,9 @@
 				GCC_PREPROCESSOR_DEFINITIONS = "STAGING=1";
 				GENERATE_INFOPLIST_FILE = YES;
 				INFOPLIST_FILE = Nos/Info.plist;
-				INFOPLIST_KEY_CFBundleDisplayName = "Nos Staging";
+				INFOPLIST_KEY_CFBundleDisplayName = "Nos Dev";
```
Comment (mplorentz) [inline Nos.xcodeproj/project.pbxproj:2866]: I think Xcode 16 added these INFOPLIST_KEYs to the project settings. They seems safe I guess.

File: CHANGELOG.md [hunk 2]
```diff
@@ -16,6 +16,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Disable autocorrect on NoteComposer when running on macOS. [#1460](https://github.com/planetary-social/nos/issues/1460)
 
 ### Internal Changes
+- Moved the database cleanup routine into a background execution task. 
+    [#1426](https://github.com/planetary-social/nos/issues/1426)
```
Comment (joshuatbrown) [inline CHANGELOG.md:20]: Looks like this somehow got into the wrong section, probably because git diffs are hard.

File: Nos/AppDelegate.swift [hunk 3]
```diff
@@ -10,6 +10,7 @@ class AppDelegate: NSObject, UIApplicationDelegate {
     @Dependency(\.currentUser) private var currentUser
     @Dependency(\.pushNotificationService) private var pushNotificationService
     @Dependency(\.analytics) private var analytics
+    @Dependency(\.persistenceController) private var persistenceController
```
Comment (joshuatbrown) [inline Nos/AppDelegate.swift:13]: 🤨

File: Nos/Controller/PersistenceController.swift [hunk 4]
```diff
@@ -160,6 +199,8 @@ final class PersistenceController {
     }
 }
 
+// MARK: - Testing & Debug
```
Comment (joshuatbrown) [inline Nos/Controller/PersistenceController.swift:202]: 💯

File: Nos/Controller/PersistenceController.swift [hunk 5]
```diff
@@ -140,6 +149,36 @@ final class PersistenceController {
         }
     }
     
+    /// Tells iOS to wake our app up in the background to run `cleanupEntities()` periodically. This tends to run 
+    /// once a day when the device is connected to power.
```
Comment (joshuatbrown) [inline Nos/Controller/PersistenceController.swift:153]: This doc comment and implementation are excellent.

File: Nos/NosApp.swift [hunk 6]
```diff
@@ -21,6 +21,7 @@ struct NosApp: App {
         // hack to fix confirmationDialog color issue
         // https://github.com/planetary-social/nos/issues/1064
         UIView.appearance(whenContainedInInstancesOf: [UIAlertController.self]).tintColor = .systemBlue
+        persistenceController.scheduleBackgroundCleanupTask()
```
Comment (joshuatbrown) [inline Nos/NosApp.swift:24]: For some reason, I'm getting a crash on this line. I took a screen shot further down the stack trace in PersistenceController:

Screenshot 2024-10-03 at 4 16 25 PM

**Review by joshuatbrown — COMMENTED** [review]
I'm not able to test due to the crash, but the code and comments all look great!

File: CHANGELOG.md (hunk 2 above)
Comment (mplorentz) [reply CHANGELOG.md:20]: Lately when I merge from `main` there are no conflicts in the changelog, even if there has been a release. Something to be aware of I guess. I'm not sure how to make it better.

**Review by joshuatbrown — APPROVED** [review]
Not crashing any more. Looks great!

File: Nos.xcodeproj/project.pbxproj (hunk 1 above)
Comment (joshuatbrown) [reply Nos.xcodeproj/project.pbxproj:2866]: @mplorentz here's where the bundle display name broke, I think. Fixed in #1620.

Comment (mplorentz) [reply Nos.xcodeproj/project.pbxproj:2866]: weird. Thanks for fixing this.

**Pull request merged**